packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/BoundedNonWhitespaceSequenceScopeHandler.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  Direction,
+  Position,
+  ScopeType,
+  TextEditor,
+  type Range,
+} from "@cursorless/common";
+import type { Target } from "../../../typings/target.types";
+import { TokenTarget } from "../../targets";
+import { BaseScopeHandler } from "./BaseScopeHandler";
+import { compareTargetScopes } from "./compareTargetScopes";
+import type { TargetScope } from "./scope.types";
+import type { ScopeHandler } from "./scopeHandler.types";
+import type { ScopeHandlerFactory } from "./ScopeHandlerFactory";
+
+export class BoundedNonWhitespaceSequenceScopeHandler extends BaseScopeHandler {
+  public readonly scopeType = { type: "boundedNonWhitespaceSequence" } as const;
+  public readonly iterationScopeType: ScopeType = {
+    type: "line",
+  } as const;
+  protected readonly isHierarchical = true;
+
+  constructor(
+    private scopeHandlerFactory: ScopeHandlerFactory,
+    _scopeType: ScopeType,
+    private languageId: string,
+  ) {
+    super();
+  }
+
+  *generateScopeCandidates(
+    editor: TextEditor,
+    position: Position,
+    direction: Direction,
+  ): Iterable<TargetScope> {
+    const paintScopeHandler = this.scopeHandlerFactory.create(
+      {
+        type: "nonWhitespaceSequence",
+      },
+      this.languageId,
+    )!;
+
+    const surroundingPairScopeHandler = this.scopeHandlerFactory.create(
+      {
+        type: "surroundingPair",
+        delimiter: "any",
+        requireStrongContainment: true,
+      },
+      this.languageId,
+    )!;
+
+    const paintScopes = paintScopeHandler.generateScopes(
+      editor,
+      position,
+      direction,
+      {
+        allowAdjacentScopes: true,
+      },
+    );
+
+    const scopes = Array.from(paintScopes)
+      .flatMap((scope) => {
+        const target = assertSingleTarget(scope);
+        const pairInteriors = getSurroundingPairInteriors(
+          surroundingPairScopeHandler,
+          target,
+        );
+        const result = [scope];
+
+        for (const pairInterior of pairInteriors) {
+          const intersection = target.contentRange.intersection(pairInterior);
+          if (intersection != null && !intersection.isEmpty) {
+            result.push(createTargetScope(editor, intersection));
+          }
+        }
+
+        return result;
+      })
+      .sort((a, b) => compareTargetScopes(direction, position, a, b));
+
+    yield* scopes;
+  }
+}
+
+function getSurroundingPairInteriors(
+  surroundingPairScopeHandler: ScopeHandler,
+  target: Target,
+): Range[] {
+  const surroundingPairScopes = surroundingPairScopeHandler.generateScopes(
+    target.editor,
+    target.contentRange.start,
+    "forward",
```

Comment:
I think you might be able to just use direction and hints from main function call?

Reply:
I don't want the surrounding pair containing the position. I want the surrounding pair containing the paint target.
Using the distal position is a good idea.

Reply:
you do want the one containing hte position. eg `aaa(bb|b)ccc`

Reply:
No I don't think so. 
eg: `|(a) (b)`
I want to yield `b` then  I need to find the surrounding pairs in each paint.